### PR TITLE
Modified Stream support functionality.

### PR DESCRIPTION
**Pull request by vchiluka5**
Modified stream support functionality
Added streams to optical_flow.cpp sample app
Modified documentation of streams wrt NVIDIA Optical Flow

<cut/>

```
force_builders=Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:18.04
```

### CODE REVIEW

**Comment by vchiluka5** [comment]
Hi @alalek 
Can you please review the above commit and if all seems well then can you please merge it to master?

**Comment by alalek** [comment]
Please take a look on build warnings from "Custom" builder.

File: modules/cudaoptflow/include/opencv2/cudaoptflow.hpp [hunk 1]
```diff
@@ -444,7 +452,9 @@ class CV_EXPORTS_W NvidiaOpticalFlow_1_0 : public NvidiaHWOpticalFlow
         bool enableTemporalHints = false,
         bool enableExternalHints = false,
         bool enableCostBuffer = false,
-        int gpuId = 0);
+        int gpuId = 0,
+        Stream inputStream = Stream::Null(),
+        Stream outputStream = Stream::Null());
```
Comment (alalek) [inline modules/cudaoptflow/include/opencv2/cudaoptflow.hpp:457]: According to other CUDA APIs this parameter is usually passed as `Stream&`.

Not sure about "const". /cc @asmorkalov

Comment (vchiluka5) [reply modules/cudaoptflow/include/opencv2/cudaoptflow.hpp:457]: > According to other CUDA APIs this parameter is usually passed as `Stream&`.
> 
> Not sure about "const". /cc @asmorkalov

When we do Stream x = Stream(), It creates a new stream and returns pointer to it right?

Comment (alalek) [reply modules/cudaoptflow/include/opencv2/cudaoptflow.hpp:457]: See [here](https://github.com/opencv/opencv/blob/4.3.0/modules/core/include/opencv2/core/cuda.hpp#L1046).

Comment (vchiluka5) [reply modules/cudaoptflow/include/opencv2/cudaoptflow.hpp]: > See [here](https://github.com/opencv/opencv/blob/4.3.0/modules/core/include/opencv2/core/cuda.hpp#L1046).

Thanks. Modified it accordingly.

**Comment by alalek** [comment]
@vchiluka5 There is build error from this patch:
```
/build/master-contrib_cuda-lin64/opencv_contrib/modules/cudaoptflow/samples/optical_flow.cpp: In function 'int main(int, const char**)':
/build/master-contrib_cuda-lin64/opencv_contrib/modules/cudaoptflow/samples/optical_flow.cpp:185:119: error: invalid initialization of non-const reference of type 'cv::cuda::Stream&' from an rvalue of type 'cv::cuda::Stream'
         NvidiaOpticalFlow_1_0::NVIDIA_OF_PERF_LEVEL::NV_OF_PERF_LEVEL_FAST, false, false, false, 0, Stream(), Stream());
                                                                                                                       ^
In file included from /build/master-contrib_cuda-lin64/opencv_contrib/modules/cudaoptflow/samples/optical_flow.cpp:7:0:
/build/master-contrib_cuda-lin64/opencv_contrib/modules/cudaoptflow/include/opencv2/cudaoptflow.hpp:447:47: note:   initializing argument 8 of 'static cv::Ptr<cv::cuda::NvidiaOpticalFlow_1_0> cv::cuda::NvidiaOpticalFlow_1_0::create(int, int, cv::cuda::NvidiaOpticalFlow_1_0::NVIDIA_OF_PERF_LEVEL, bool, bool, bool, int, cv::cuda::Stream&, cv::cuda::Stream&)'
     CV_WRAP static Ptr<NvidiaOpticalFlow_1_0> create(
                                               ^
```

Please take a look. Perhaps "const reference" is required in API.

**Comment by tomoaki0705** [comment]
please have a look in #2540 
that patch worked for me

**Comment by vchiluka5** [comment]
> please have a look in #2540
> that patch worked for me

@tomoaki0705  we need to pass a valid stream and not NULL() stream. 
@alalek  Its strange since i was able to build and run the app as well on windows. Also builds didnt gave any error right before merging PR?
Let me fix it. I still dont see error on my windows machine. Il setup a linux machine and try it out. Thanks for pointing it out.

**Pull request merged**